### PR TITLE
Main version code before experiments

### PR DESCRIPTION
**Pull request by Simontwice**
# Description
Pushes the code to a remote branch before running an experiment remotely.
NOTE: the state of the IDE/commits/uncommited changes are exactly the same before and after this code is executed. Automagic!!!

# Checklist
Answer each question with "yes", and provide an explanation if the answer isn't what is expected.

1. Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - Answer: y
2. Are all variables set with logical defaults that are backward compatible?
   - Answer: y
3. Can you provide a link to a Neptune experiment verifying the functionality of this PR?
   - Answer: no point
4. Have you successfully executed the linter and tests?
   - Answer: y
5. Are all functions concise and don't require splitting?
   - Answer: y
6. Is there documentation for arguments and complex logic?
   - Answer: y
7. Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - Answer: y
8. Are all functions placed in appropriate files?
   - Answer: y
9. Is the PR name meaningful and descriptive?
   - Answer: y
10. Is PR description provided, or unnecessary?
    - Answer: y
  
Other questions:

11. Have you not made changes to the requirements or anything related to the Singularity image? If you have, did you generate and upload a new image to the clusters?
    - Answer: n
12. Is the change significant enough to notify all individuals working with this repository?
    - Answer: y
13. Is there a need for a second reviewer?
    - Answer: n

### CODE REVIEW

File: lizrd/support/code_copying_support.py
Comment (sebastianjaszczur) [inline lizrd/support/code_copying_support.py:1]: I think that '_support' in filename is unnecessary if we have support/ in path?

**Review by sebastianjaszczur — COMMENTED** [review]
LGTM, ale nie daję approve - chciałbym jednak, żeby ktoś jeszcze sprawdził kod, ale przede wszystkim spróbował to odpalić czy działa. A generalnie, Szymon, czy to działa na każdym serwerze? Entropy/athena/dgx?

**Comment by Simontwice** [comment]
> LGTM, ale nie daję approve - chciałbym jednak, żeby ktoś jeszcze sprawdził kod, ale przede wszystkim spróbował to odpalić czy działa. A generalnie, Szymon, czy to działa na każdym serwerze? Entropy/athena/dgx?

Działa na wszystkim, bo nigdy nie dotyka serwera - to się dzieje lokalnie; przed wysłaniem
Czekam na testerów

**Comment by sebastianjaszczur** [comment]
How is testing going?

**Comment by Simontwice** [comment]
I cannot get anyone to do a review.

File: lizrd/support/sync_and_version.py [hunk 1]
```diff
@@ -2,10 +2,11 @@
 from fabric import Connection
 from argparse import ArgumentParser
 from git import Repo
-from lizrd.support.code_versioning_support import find_git_root
+from lizrd.support.code_versioning import find_git_root, version_code
+from lizrd.support.misc import generate_random_string
 
 
-def cd_to_root_dir():
+def go_to_llm_random():
```
Comment (crewtool) [inline lizrd/support/sync_and_version.py:9]: imo cd_to_root_dir was better

**Review by crewtool — APPROVED** [review]

File: lizrd/support/code_versioning.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
+import os
+import subprocess
+from typing import Union
+
+from git import Repo, GitCommandError
+
+
+class CodeVersioningDaemon:
```
[the remark this answers is not in the document]
Comment (Simontwice) [reply lizrd/support/code_versioning.py:8]: d

File: lizrd/support/code_copying.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import shutil
+
+
+def copy_code(newdir_name):
+    # Copy code
+    root_dir = os.getcwd()
+    assert root_dir.endswith(
+        "llm-random"
+    ), "You need to be in the llm-random directory to copy code."
+    newdir_path = f"{os.path.dirname(root_dir)}/sparsity_code_cemetery/{newdir_name}"
```
[the remark this answers is not in the document]
Comment (Simontwice) [reply lizrd/support/code_copying.py:11]: d

File: lizrd/support/code_copying.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import os
+import shutil
+
+
+def copy_code(newdir_name):
+    # Copy code
+    root_dir = os.getcwd()
+    assert root_dir.endswith(
```
[the remark this answers is not in the document]
Comment (Simontwice) [reply lizrd/support/code_copying.py:8]: d

File: lizrd/support/code_versioning.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                self.repo.iter_commits(f"origin/{self.original_branch}..HEAD")
+            )
+            if len(commits_behind) > 0:
+                raise Exception(
+                    f"Either branch is does not track any remote branch [you haven't pushed anything yet] OR unpushed commits have been detected. Either way, push first. Aborting..."
+                )
+
+            self.check_if_cemetery_exists()
+            self.stash_if_necessary()
+            self.revert_status = 1
```
[the remark this answers is not in the document]
Comment (Simontwice) [reply lizrd/support/code_versioning.py:46]: d

File: lizrd/support/code_versioning.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+            )
+
+        except GitCommandError:
+            self.handle_failure()
+            raise Exception("Failed to version code. Aborting...")
+
+    def handle_failure(self):
+        if self.revert_status == 0:
+            pass
+        elif self.revert_status == 1:
```
[the remark this answers is not in the document]
Comment (Simontwice) [reply lizrd/support/code_versioning.py:67]: d

File: lizrd/support/code_versioning.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    def reset_to_original_branch_and_commit(self):
+        self.repo.git.checkout(self.original_branch, "-f")
+        self.repo.head.reset(
+            self.original_branch_commit_hash,
+            index=True,
+            working_tree=True,
+        )
+        self.unstash_if_necessary()
+
+    def check_if_cemetery_exists(self):
```
[the remark this answers is not in the document]
Comment (Simontwice) [reply lizrd/support/code_versioning.py:87]: d

File: lizrd/support/code_copying_support.py
Comment (Simontwice) [reply lizrd/support/code_copying_support.py:1]: fair, changed

**Pull request merged**